### PR TITLE
ReactiveAerospikeTemplate implementation (find-related methods)

### PR DESCRIPTION
**Pull request by igor-ermolenko**
ReactiveAerospikeTemplate implementation (find-related methods)

### CODE REVIEW

**Comment by igor-ermolenko** [comment]
@Aloren @BrianNichols could you please review?

File: src/main/java/org/springframework/data/aerospike/core/ReactiveAerospikeTemplate.java [hunk 1]
```diff
@@ -76,15 +81,66 @@ public ReactiveAerospikeTemplate(AerospikeClient client,
...
+        Assert.notNull(type, "Type must not be null!");
 
-        AerospikePersistentEntity<?> entity = mappingContext.getPersistentEntity(targetType);
+        AerospikePersistentEntity<?> entity = mappingContext.getRequiredPersistentEntity(type);
         Key key = getKey(id, entity);
 
-        return reactorClient.get(key)
-                .map(keyRecord -> mapToEntityOptional(keyRecord.key, targetType, keyRecord.record));
+        if (entity.isTouchOnRead()) {
+            Assert.state(!entity.hasExpirationProperty(), "Touch on read is not supported for expiration property");
```
Comment (Aloren) [inline src/main/java/org/springframework/data/aerospike/core/ReactiveAerospikeTemplate.java]: Touch on read is not supported for expiration property ->
Touch on read is not supported for entity without expiration property

File: src/test/java/org/springframework/data/aerospike/BaseIntegrationTests.java [hunk 2]
```diff
@@ -53,6 +53,9 @@ protected static String nextId() {
     protected void cleanDb() {
         template.delete(Person.class);
         template.delete(SampleClasses.VersionedClass.class);
+        template.delete(SampleClasses.DocumentWithExpiration.class);
```
Comment (Aloren) [inline src/test/java/org/springframework/data/aerospike/BaseIntegrationTests.java]: if you generate unique id for each test -- there is no need to cleanup after test i suppose

File: src/test/java/org/springframework/data/aerospike/core/reactive/BaseReactiveAerospikeTemplateTests.java [hunk 3]
```diff
@@ -15,11 +31,33 @@
...
         cleanDb();
     }
+
+    private WritePolicy getWritePolicy() {
+        WritePolicy policy = new WritePolicy();
+        policy.sendKey = true;
+        return policy;
+    }
+
+    <T> Query createQueryForMethodWithArgs(String methodName, Object... args) {
```
Comment (Aloren) [inline src/test/java/org/springframework/data/aerospike/core/reactive/BaseReactiveAerospikeTemplateTests.java]: as i remember, this method was already present in some other class. makes sense to extract to `AerospikeTemplateTestUtils` class

File: src/test/java/org/springframework/data/aerospike/core/reactive/ReactiveAerospikeTemplateFindRelatedTests.java [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+import static org.junit.Assert.*;
+import static org.springframework.data.aerospike.SampleClasses.EXPIRATION_ONE_MINUTE;
+
+/**
+ * Tests for find related methods in {@link ReactiveAerospikeTemplate}.
+ *
+ * @author Igor Ermolenko
+ */
+public class ReactiveAerospikeTemplateFindRelatedTests extends BaseReactiveAerospikeTemplateTests {
```
Comment (Aloren) [inline src/test/java/org/springframework/data/aerospike/core/reactive/ReactiveAerospikeTemplateFindRelatedTests.java]: nice idea

File: src/test/java/org/springframework/data/aerospike/core/reactive/ReactiveAerospikeTemplateFindRelatedTests.java [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+    static private final String SET_NAME_PERSON = "Person";
+
+    @Test
+    public void testFindById() {
```
Comment (Aloren) [inline src/test/java/org/springframework/data/aerospike/core/reactive/ReactiveAerospikeTemplateFindRelatedTests.java]: if you don't mind -- please rename test methods to be named in the same style.
`findById_shouldReturnNullForNonExistingKey ` -- ok
`testFindById` -- bad

File: src/test/java/org/springframework/data/aerospike/core/reactive/ReactiveAerospikeTemplateSaveRelatedTests.java [hunk 6]
```diff
@@ -34,7 +34,7 @@ public void shouldSaveAndSetVersion() {
         reactiveTemplate.save(first).block();
 
         assertThat(first.version).isEqualTo(1);
-        assertThat(template.findById(id, VersionedClass.class).version).isEqualTo(1);
+        assertThat(findById(id, VersionedClass.class).version).isEqualTo(1);
```
Comment (Aloren) [inline src/test/java/org/springframework/data/aerospike/core/reactive/ReactiveAerospikeTemplateSaveRelatedTests.java]: https://github.com/reactor/reactor-core/blob/master/reactor-test/src/test/java/reactor/test/StepVerifierTests.java
do we need to use this approach instead?

File: src/main/java/org/springframework/data/aerospike/core/ReactiveAerospikeTemplate.java (hunk 1 above)
Comment (igor-ermolenko) [reply src/main/java/org/springframework/data/aerospike/core/ReactiveAerospikeTemplate.java]: Fixed

File: src/test/java/org/springframework/data/aerospike/core/reactive/ReactiveAerospikeTemplateFindRelatedTests.java (hunk 4 above)
Comment (igor-ermolenko) [reply src/test/java/org/springframework/data/aerospike/core/reactive/ReactiveAerospikeTemplateFindRelatedTests.java]: Thanks

File: src/test/java/org/springframework/data/aerospike/core/reactive/ReactiveAerospikeTemplateFindRelatedTests.java (hunk 5 above)
Comment (igor-ermolenko) [reply src/test/java/org/springframework/data/aerospike/core/reactive/ReactiveAerospikeTemplateFindRelatedTests.java]: I don't mind. Tests renamed.

File: src/test/java/org/springframework/data/aerospike/core/reactive/BaseReactiveAerospikeTemplateTests.java (hunk 3 above)
Comment (igor-ermolenko) [reply src/test/java/org/springframework/data/aerospike/core/reactive/BaseReactiveAerospikeTemplateTests.java]: Done

File: src/test/java/org/springframework/data/aerospike/BaseIntegrationTests.java (hunk 2 above)
Comment (igor-ermolenko) [reply src/test/java/org/springframework/data/aerospike/BaseIntegrationTests.java]: fixed

**Comment by Aloren** [comment]
lgtm

**Comment by igor-ermolenko** [comment]
@BrianNichols could you please merge this PR into master if everything suits? Release not needed. Thank you!

File: src/test/java/org/springframework/data/aerospike/core/reactive/ReactiveAerospikeTemplateSaveRelatedTests.java (hunk 6 above)
Comment (igor-ermolenko) [reply src/test/java/org/springframework/data/aerospike/core/reactive/ReactiveAerospikeTemplateSaveRelatedTests.java]: I'll investigate this point.

**Comment by BrianNichols** [comment]
Done.

**Comment by igor-ermolenko** [comment]
Thanks!

**Pull request merged**